cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageWriter.java

```diff
@@ -109,49 +115,118 @@ public void close() throws IOException {
-            long contentLength = get.getContentLength();
-            Date lastModified = get.getLastModified();
-
-            // create entry (metadata) to be put to archive stream
-            log.debug("next package entry: " + packageItem.getRelativePath() + "," + contentLength + "," + lastModified);
-            ArchiveEntry e = createEntry(packageItem.getRelativePath(), contentLength, lastModified);
-
-            // put archive entry to stream
-            aout.putArchiveEntry(e);
 
-            // headers for entry have been written, body has not,
-            // so consider this entry 'open'
-            openEntry = true;
-
-            // Copy the get InputStream to the package OutputStream
-            // this is 'writing' the content of the file
-            InputStream getIOStream = get.getInputStream();
-            MultiBufferIO multiBufferIO = new MultiBufferIO();
-            multiBufferIO.copy(getIOStream, aout);
+            if (packageURL.getProtocol().equals("file")) {
+                writeFileContentEntry(packageURL, packageItem.getRelativePath());
+            } else {
+                writeHttpContentEntry(packageURL, packageItem.getRelativePath());
+            }
 
         } finally {
             if (openEntry) {
                 aout.closeArchiveEntry();
             }
         }
     }
+
+
+    /**
+     * Set up header for ArchiveEntry in the current ArchiveOutputStream.
+     * @param relativePath - relative path of item in the package file
+     * @param contentLength - (long) - bytes to be written
+     * @param lastModified - (Date) - last modified date of item
+     * @return
+     * @throws IOException
+     */
+    private boolean setArchiveEntryHeader(String relativePath, long contentLength, Date lastModified)
+        throws IOException {
+
+        // create entry (metadata) to be put to archive stream
+        log.debug("next package entry: " + relativePath + "," + contentLength + "," + lastModified);
+        ArchiveEntry e = createEntry(relativePath, contentLength, lastModified);
+
+        // put archive entry to stream
+        aout.putArchiveEntry(e);
+
+        // headers for entry have been written, body has not,
+        // so consider this entry 'open'
+        return true;
+    }
+
+    /**
+     * Write the content of a file:// URL to the current ArchiveOutputStream
+     * @param fileURL - URL of file item being written to
+     * @param relativePath - relative path of file inside package beingw written.
+     * @throws IOException
+     */
+    private void writeFileContentEntry(URL fileURL, String relativePath) throws IOException {
+        Path filePath = FileSystems.getDefault().getPath(fileURL.getPath());
+        log.debug("filePath from fileURL: " + filePath.toString() + ": " + fileURL.toString());
+
+        long contentLength = Files.size(filePath);
+        FileTime lastMod = (FileTime)Files.getAttribute(filePath, "lastModifiedTime");
```

Comment:
There is a 'getLastModifiedTime' method that's probably better to use.

Reply:
ok